## **Importing the libraries**

In [91]:
import pandas as pd 
import numpy as np 
import seaborn as sns 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

## **Loading The Dataset**

In [92]:
df = pd.read_csv('Housing_Area.csv')
df.head(10)

Unnamed: 0    price  lotsize  bedrooms  bathrms  stories driveway recroom  \
0           1  42000.0     5850         3        1        2      yes      no   
1           2  38500.0     4000         2        1        1      yes      no   
2           3  49500.0     3060         3        1        1      yes      no   
3           4  60500.0     6650         3        1        2      yes     yes   
4           5  61000.0     6360         2        1        1      yes      no   
5           6  66000.0     4160         3        1        1      yes     yes   
6           7  66000.0     3880         3        2        2      yes      no   
7           8  69000.0     4160         3        1        3      yes      no   
8           9  83800.0     4800         3        1        1      yes     yes   
9          10  88500.0     5500         3        2        4      yes     yes   

  fullbase gashw airco  garagepl prefarea  
0      yes    no    no         1       no  
1       no    no    no         0       no  
2       no    no    no         0       no  
3       no    no    no         0       no  
4       no    no    no         0       no  
5      yes    no   yes         0       no  
6      yes    no    no         2       no  
7       no    no    no         0       no  
8      yes    no    no         0       no  
9       no    no   yes         1       no

In [93]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)

In [94]:
lb = LabelEncoder()
st = StandardScaler()

# Label Encoding

In [95]:
df['driveway'] = lb.fit_transform(df['driveway'])
df['recroom'] = lb.fit_transform(df['recroom'])
df['fullbase'] = lb.fit_transform(df['fullbase'])
df['gashw'] = lb.fit_transform(df['gashw'])
df['airco'] = lb.fit_transform(df['airco'])
df['prefarea'] = lb.fit_transform(df['prefarea'])

In [96]:
df

price  lotsize  bedrooms  bathrms  stories  driveway  recroom  \
0     42000.0     5850         3        1        2         1        0   
1     38500.0     4000         2        1        1         1        0   
2     49500.0     3060         3        1        1         1        0   
3     60500.0     6650         3        1        2         1        1   
4     61000.0     6360         2        1        1         1        0   
..        ...      ...       ...      ...      ...       ...      ...   
541   91500.0     4800         3        2        4         1        1   
542   94000.0     6000         3        2        4         1        0   
543  103000.0     6000         3        2        4         1        1   
544  105000.0     6000         3        2        2         1        1   
545  105000.0     6000         3        1        2         1        0   

     fullbase  gashw  airco  garagepl  prefarea  
0           1      0      0         1         0  
1           0      0      0         0         0  
2           0      0      0         0         0  
3           0      0      0         0         0  
4           0      0      0         0         0  
..        ...    ...    ...       ...       ...  
541         0      0      1         0         0  
542         0      0      1         0         0  
543         0      0      1         1         0  
544         0      0      1         1         0  
545         0      0      1         1         0  

[546 rows x 12 columns]

In [97]:
data_copy = df.copy()

# Feature scaling

In [98]:
df[['lotsize']] = st.fit_transform(df[['lotsize']])
df[['bedrooms']] = st.fit_transform(df[['bedrooms']])
df[['bathrms']] = st.fit_transform(df[['bathrms']])
df[['stories']] = st.fit_transform(df[['stories']])
df[['recroom']] = st.fit_transform(df[['recroom']])
df[['fullbase']] = st.fit_transform(df[['fullbase']])
df[['gashw']] = st.fit_transform(df[['gashw']])
df[['airco']] = st.fit_transform(df[['airco']])
df[['price']] = st.fit_transform(df[['price']])
df[['driveway']] = st.fit_transform(df[['driveway']])
df[['garagepl']] = st.fit_transform(df[['garagepl']])

In [99]:
df

price   lotsize  bedrooms   bathrms   stories  driveway   recroom  \
0   -0.979136  0.323028  0.047235 -0.569495  0.221704   0.40519 -0.464796   
1   -1.110329 -0.531013 -1.310147 -0.569495 -0.931157   0.40519 -0.464796   
2   -0.698008 -0.964958  0.047235 -0.569495 -0.931157   0.40519 -0.464796   
3   -0.285686  0.692343  0.047235 -0.569495  0.221704   0.40519  2.151480   
4   -0.266944  0.558466 -1.310147 -0.569495 -0.931157   0.40519 -0.464796   
..        ...       ...       ...       ...       ...       ...       ...   
541  0.876311 -0.161698  0.047235  1.423737  2.527427   0.40519  2.151480   
542  0.970020  0.392275  0.047235  1.423737  2.527427   0.40519 -0.464796   
543  1.307374  0.392275  0.047235  1.423737  2.527427   0.40519  2.151480   
544  1.382342  0.392275  0.047235  1.423737  0.221704   0.40519  2.151480   
545  1.382342  0.392275  0.047235 -0.569495  0.221704   0.40519 -0.464796   

     fullbase     gashw     airco  garagepl  prefarea  
0    1.363319 -0.219054 -0.681034  0.357567         0  
1   -0.733504 -0.219054 -0.681034 -0.804525         0  
2   -0.733504 -0.219054 -0.681034 -0.804525         0  
3   -0.733504 -0.219054 -0.681034 -0.804525         0  
4   -0.733504 -0.219054 -0.681034 -0.804525         0  
..        ...       ...       ...       ...       ...  
541 -0.733504 -0.219054  1.468356 -0.804525         0  
542 -0.733504 -0.219054  1.468356 -0.804525         0  
543 -0.733504 -0.219054  1.468356  0.357567         0  
544 -0.733504 -0.219054  1.468356  0.357567         0  
545 -0.733504 -0.219054  1.468356  0.357567         0  

[546 rows x 12 columns]

In [100]:
X = df.drop(['prefarea'], axis=1)

In [101]:
X

price   lotsize  bedrooms   bathrms   stories  driveway   recroom  \
0   -0.979136  0.323028  0.047235 -0.569495  0.221704   0.40519 -0.464796   
1   -1.110329 -0.531013 -1.310147 -0.569495 -0.931157   0.40519 -0.464796   
2   -0.698008 -0.964958  0.047235 -0.569495 -0.931157   0.40519 -0.464796   
3   -0.285686  0.692343  0.047235 -0.569495  0.221704   0.40519  2.151480   
4   -0.266944  0.558466 -1.310147 -0.569495 -0.931157   0.40519 -0.464796   
..        ...       ...       ...       ...       ...       ...       ...   
541  0.876311 -0.161698  0.047235  1.423737  2.527427   0.40519  2.151480   
542  0.970020  0.392275  0.047235  1.423737  2.527427   0.40519 -0.464796   
543  1.307374  0.392275  0.047235  1.423737  2.527427   0.40519  2.151480   
544  1.382342  0.392275  0.047235  1.423737  0.221704   0.40519  2.151480   
545  1.382342  0.392275  0.047235 -0.569495  0.221704   0.40519 -0.464796   

     fullbase     gashw     airco  garagepl  
0    1.363319 -0.219054 -0.681034  0.357567  
1   -0.733504 -0.219054 -0.681034 -0.804525  
2   -0.733504 -0.219054 -0.681034 -0.804525  
3   -0.733504 -0.219054 -0.681034 -0.804525  
4   -0.733504 -0.219054 -0.681034 -0.804525  
..        ...       ...       ...       ...  
541 -0.733504 -0.219054  1.468356 -0.804525  
542 -0.733504 -0.219054  1.468356 -0.804525  
543 -0.733504 -0.219054  1.468356  0.357567  
544 -0.733504 -0.219054  1.468356  0.357567  
545 -0.733504 -0.219054  1.468356  0.357567  

[546 rows x 11 columns]

In [102]:
Y= df[['prefarea']]

In [103]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [104]:
model = LogisticRegression()

In [105]:
model.fit(x_train, y_train)

C:\Users\Ali Hasnain\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

In [106]:
y_pred= model.predict(x_test)

In [107]:
y_pred

array([0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

## **Testing score**

In [108]:
model.score(x_test, y_test)

0.8636363636363636

## **Training score**

In [109]:
model.score(x_train, y_train)

0.7935779816513762

## **Overall accuracy**

In [110]:
from sklearn.metrics import accuracy_score
print("Accuracy : ", accuracy_score(y_test, y_pred))

Accuracy :  0.8636363636363636


## **Wild Test**

In [111]:
import statsmodels.formula.api as smf 
mlr = smf.ols(formula='prefarea ~ price + lotsize + bedrooms + bathrms + stories + driveway + recroom + fullbase + gashw + airco + garagepl', data=df).fit() 

In [112]:
waldtest = mlr.wald_test(r_matrix="price", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[31.51187205]]), p=3.189602352956903e-08, df_denom=534, df_num=1>

In [113]:
waldtest = mlr.wald_test(r_matrix="lotsize", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[1.36939787]]), p=0.24243735404353745, df_denom=534, df_num=1>

In [114]:
waldtest = mlr.wald_test(r_matrix="bedrooms", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.00429222]]), p=0.9477884658566821, df_denom=534, df_num=1>

In [115]:
waldtest = mlr.wald_test(r_matrix="bathrms", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[6.54675037]]), p=0.010782228120118921, df_denom=534, df_num=1>

In [116]:
waldtest = mlr.wald_test(r_matrix="stories", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.86182355]]), p=0.35364830940450864, df_denom=534, df_num=1>

In [117]:
waldtest = mlr.wald_test(r_matrix="driveway", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[4.56352722]]), p=0.03311375815361842, df_denom=534, df_num=1>

In [118]:
waldtest = mlr.wald_test(r_matrix="recroom", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.15364583]]), p=0.69523172569154, df_denom=534, df_num=1>

In [119]:
waldtest = mlr.wald_test(r_matrix="fullbase", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[12.01808638]]), p=0.0005693093118276944, df_denom=534, df_num=1>

In [120]:
waldtest = mlr.wald_test(r_matrix="gashw", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[4.57874453]]), p=0.03282349094653312, df_denom=534, df_num=1>

In [121]:
waldtest = mlr.wald_test(r_matrix="airco", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[1.29053751]]), p=0.25645849319995717, df_denom=534, df_num=1>

In [122]:
waldtest = mlr.wald_test(r_matrix="garagepl", use_f=True, scalar=False)
waldtest

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[1.95638432]]), p=0.162481044184708, df_denom=534, df_num=1>

## **As from wild test result , the p value (parameter value) for bedrooms in data has highest contribution  p=0.947788465856682**

---

## **Likelihood test**

In [123]:
import statsmodels.api as sm
full_model = sm.OLS(y_train, x_train).fit()
full_ll = full_model.llf
full_ll

-278.46507791398426

## **Confusion Matrix**

In [124]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

print ("Confusion Matrix : \n", cm)

Confusion Matrix : 
 [[89  1]
 [14  6]]


## **Here 89 + 6 values are correctly predicit by our model and 14 + 1 values are predicit wrong**

---

## **Overall summary**

In [125]:
import statsmodels.api as sm
logit_model=sm.Logit(Y,X)
result=logit_model.fit()
print(result.summary2())

Optimization terminated successfully.
         Current function value: 0.623868
         Iterations 5
                        Results: Logit
Model:              Logit            Pseudo R-squared: -0.146  
Dependent Variable: prefarea         AIC:              703.2634
Date:               2022-08-10 02:44 BIC:              750.5922
No. Observations:   546              Log-Likelihood:   -340.63 
Df Model:           10               LL-Null:          -297.34 
Df Residuals:       535              LLR p-value:      1.0000  
Converged:          1.0000           Scale:            1.0000  
No. Iterations:     5.0000                                     
----------------------------------------------------------------
             Coef.   Std.Err.     z     P>|z|    [0.025   0.975]
----------------------------------------------------------------
price        0.8030    0.1768   4.5427  0.0000   0.4565   1.1494
lotsize      0.0966    0.1166   0.8287  0.4073  -0.1319   0.3251
bedrooms     0.0118   

In [126]:
import statsmodels.api as sm
logit_model=sm.Logit(y_train,x_train)
result=logit_model.fit()
print(result.summary2())

Optimization terminated successfully.
         Current function value: 0.622333
         Iterations 5
                        Results: Logit
Model:              Logit            Pseudo R-squared: -0.112  
Dependent Variable: prefarea         AIC:              564.6743
Date:               2022-08-10 02:44 BIC:              609.5283
No. Observations:   436              Log-Likelihood:   -271.34 
Df Model:           10               LL-Null:          -244.07 
Df Residuals:       425              LLR p-value:      1.0000  
Converged:          1.0000           Scale:            1.0000  
No. Iterations:     5.0000                                     
----------------------------------------------------------------
             Coef.   Std.Err.     z     P>|z|    [0.025   0.975]
----------------------------------------------------------------
price        0.6683    0.1969   3.3947  0.0007   0.2824   1.0541
lotsize      0.1822    0.1327   1.3729  0.1698  -0.0779   0.4422
bedrooms    -0.0109   

In [127]:
import statsmodels.api as sm
logit_model=sm.Logit(y_test,x_test)
result=logit_model.fit()
print(result.summary2())

Optimization terminated successfully.
         Current function value: 0.550901
         Iterations 6
                        Results: Logit
Model:              Logit            Pseudo R-squared: -0.162  
Dependent Variable: prefarea         AIC:              143.1981
Date:               2022-08-10 02:44 BIC:              172.9034
No. Observations:   110              Log-Likelihood:   -60.599 
Df Model:           10               LL-Null:          -52.155 
Df Residuals:       99               LLR p-value:      1.0000  
Converged:          1.0000           Scale:            1.0000  
No. Iterations:     6.0000                                     
----------------------------------------------------------------
             Coef.   Std.Err.     z     P>|z|    [0.025   0.975]
----------------------------------------------------------------
price        1.3946    0.4720   2.9546  0.0031   0.4695   2.3198
lotsize     -0.1720    0.2985  -0.5763  0.5644  -0.7570   0.4130
bedrooms     0.1769   

In [128]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.99      0.92        90
           1       0.86      0.30      0.44        20

    accuracy                           0.86       110
   macro avg       0.86      0.64      0.68       110
weighted avg       0.86      0.86      0.84       110



---

## **Chi Square Test**

In [129]:
from scipy.stats import chi2_contingency
from scipy.stats import chi2

In [130]:
data_copy

price  lotsize  bedrooms  bathrms  stories  driveway  recroom  \
0     42000.0     5850         3        1        2         1        0   
1     38500.0     4000         2        1        1         1        0   
2     49500.0     3060         3        1        1         1        0   
3     60500.0     6650         3        1        2         1        1   
4     61000.0     6360         2        1        1         1        0   
..        ...      ...       ...      ...      ...       ...      ...   
541   91500.0     4800         3        2        4         1        1   
542   94000.0     6000         3        2        4         1        0   
543  103000.0     6000         3        2        4         1        1   
544  105000.0     6000         3        2        2         1        1   
545  105000.0     6000         3        1        2         1        0   

     fullbase  gashw  airco  garagepl  prefarea  
0           1      0      0         1         0  
1           0      0      0         0         0  
2           0      0      0         0         0  
3           0      0      0         0         0  
4           0      0      0         0         0  
..        ...    ...    ...       ...       ...  
541         0      0      1         0         0  
542         0      0      1         0         0  
543         0      0      1         1         0  
544         0      0      1         1         0  
545         0      0      1         1         0  

[546 rows x 12 columns]

In [133]:
stat, p, dof, expected = chi2_contingency(data_copy[['price', 'prefarea', 'lotsize', 
                                                     'bedrooms', 'bathrms', 'stories', 'driveway', 'recroom', 'fullbase', 'gashw', 'airco', 'garagepl']])

In [136]:
prob = 0.85
chi = chi2.ppf(prob, dof)
chi

6108.531507734429

In [137]:
if abs(stat) >= chi:
    	print('reject Null hypothesis')
else:
	print('fail to reject')

reject Null hypothesis
